File: 14-TypeScript/Grey Multi-Tenant/packages/grey-kotlin/src/main/kotlin/com/grey/sdk/grpc/UserGrpcService.kt
// =============================================================================
// Grey Multi-Tenant SDK - Kotlin
// grpc/UserGrpcService.kt
//
// gRPC service stub for user operations.
// =============================================================================

package com.grey.sdk.grpc

import com.grey.sdk.config.GreyOptions
import com.grey.sdk.error.GreyError
import com.grey.sdk.error.GreyResult
import io.grpc.StatusException
import io.grpc.StatusRuntimeException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * User data model.
 */
data class UserResponse(
    val id: String,
    val email: String,
    val name: String?,
    val avatarUrl: String?,
    val createdAt: Long,
    val updatedAt: Long
)

/**
 * gRPC service stub for user operations.
 * Wraps the Grey User gRPC service.
 */
class UserGrpcService(
    private val channel: GreyChannel,
    private val options: GreyOptions
) {
    /**
     * Get the current user.
     *
     * @param accessToken Current access token
     * @return GreyResult containing User or error
     */
    suspend fun getUser(accessToken: String): GreyResult<UserResponse> {
        return withContext(Dispatchers.IO) {
            try {
                val grpcChannel = channel.getChannel()
                
                // Note: In production, call actual gRPC stub with auth metadata
                // val metadata = channel.createAuthMetadata(accessToken)
                // val stub = UserServiceGrpcKt.UserServiceCoroutineStub(grpcChannel)
                //     .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata))
                // val response = stub.getUser(getUserRequest {})
                
                // Simulated response - replace with actual gRPC call
                GreyResult.success(
                    UserResponse(
                        id = "", // From gRPC response
                        email = "",
                        name = null,
                        avatarUrl = null,
                        createdAt = 0,
                        updatedAt = 0
                    )
                )
            } catch (e: StatusException) {
                GreyResult.failure(GreyError.fromGrpcStatus(e))
            } catch (e: StatusRuntimeException) {
                GreyResult.failure(GreyError.fromGrpcStatusRuntime(e))
            } catch (e: Throwable) {
                GreyResult.failure(GreyError.fromException(e))
            }
        }
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/GRPC/UserGRPCService.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// GRPC/UserGRPCService.swift
//
// gRPC service stub for user operations.
// =============================================================================

import Foundation
import GRPC

// MARK: - Response Models

/// User response data.
public struct UserResponse: Sendable {
    public let id: String
    public let email: String
    public let name: String?
    public let avatarUrl: String?
    public let createdAt: Int64
    public let updatedAt: Int64
    
    public init(
        id: String,
        email: String,
        name: String?,
        avatarUrl: String?,
        createdAt: Int64,
        updatedAt: Int64
    ) {
        self.id = id
        self.email = email
        self.name = name
        self.avatarUrl = avatarUrl
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

// MARK: - Service

/// gRPC service stub for user operations.
/// Wraps the Grey User gRPC service.
public final class UserGRPCService: Sendable {
    
    private let channel: GreyChannel
    
    // MARK: - Initialization
    
    public init(channel: GreyChannel) {
        self.channel = channel
    }
    
    // MARK: - Operations
    
    /// Get the current user.
    /// - Parameter accessToken: Current access token
    /// - Returns: UserResponse on success
    /// - Throws: GreyError on failure
    public func getUser(accessToken: String) async throws -> UserResponse {
        do {
            let grpcChannel = channel.getChannel()
            let options = channel.callOptions(accessToken: accessToken)
            
            // Note: In production, call actual gRPC stub
            // let client = User_UserServiceAsyncClient(channel: grpcChannel)
            // let response = try await client.getUser(User_GetUserRequest(), callOptions: options)
            
            return UserResponse(
                id: "",
                email: "",
                name: nil,
                avatarUrl: nil,
                createdAt: 0,
                updatedAt: 0
            )
        } catch let error as GRPCStatus {
            throw GreyError.fromGRPCStatus(error)
        } catch {
            throw GreyError.fromError(error)
        }
    }
}
